### PR TITLE
Develop multi-modal basedataset and add the StrokePatientMIdataset based on it 

### PR DESCRIPTION
**Pull request by arrogant-R**

### CODE REVIEW

**Comment by arrogant-R** [comment]
Fix some transforms and add constant of strokesmi]

File: torcheeg/transforms/numpy/downsample.py [hunk 1]
```diff
@@ -60,4 +61,53 @@ def repr_body(self) -> Dict:
...
+    .. automethod:: __call__
+    '''
+    def __init__(self,origin:int, target_sampling_rate:int, apply_to_baseline=False):
+        super(SetSamplingRate, self).__init__(apply_to_baseline=apply_to_baseline)
+        self.original_rate = origin
+        self.new_rate = target_sampling_rate
+
+
+    def apply(self, eeg, **kwargs) -> any:
+        assert len(eeg.shape) == 2, 'Make sure the EEG in 2D shape.'
```
Comment (tczhangzhi) [inline torcheeg/transforms/numpy/downsample.py:96]: The EEG signals to be downsampled do not have to be 2D. A typical example is time series signals of different frequency bands represented by [frequency band, electrode, data point].

File: torcheeg/transforms/any/baseline.py [hunk 2]
```diff
@@ -47,9 +47,42 @@ def apply(self, eeg: any, **kwargs) -> any:
...
+    def __init__(self):
+        super(BaselineCorrection, self).__init__(apply_to_baseline=False)
+        
+
+    def __call__(self, *args, eeg: any, baseline= None, **kwargs) :
+        return super().__call__(*args, eeg=eeg, baseline=baseline, **kwargs)
+
+
+    def apply(self, eeg, **kwargs) -> any:
+         assert len(eeg.shape) == 2 and  len(kwargs['baseline'].shape) ==2, "Make sure EEG and its baseline in 2D shape. "
```
Comment (tczhangzhi) [inline torcheeg/transforms/any/baseline.py:78]: This implementation does not actually meet expectations. In many works, baseline removal is applied to features, corresponding to the average of high-order tensors along the segment dimension, e.g., [1]. This is also why we did not directly use the implementation from EEGLab.

Yang Y, Wu Q, Qiu M, et al. Emotion recognition from multi-channel EEG through parallel convolutional recurrent neural network[C]//2018 international joint conference on neural networks (IJCNN). IEEE, 2018: 1-7.

File: torcheeg/datasets/module/motor_imagery/strokes.py [hunk 3]
```diff
@@ -0,0 +1,313 @@
...
+                 num_worker: int = 0,
+                 verbose: bool = True,
+                ):
+        if io_path is None:
+            io_path = get_random_dir_path(dir_prefix='datasets')
+
+        self.subjects_info = pd.read_csv(os.path.join(root_path,
+                                                      'participants.tsv'),
+                                         sep='\t')
+        self.electodes_info = pd.read_csv(os.path.join(
```
Comment (tczhangzhi) [inline torcheeg/datasets/module/motor_imagery/strokes.py:87]: The electrode information should be defined as a constant. The event information should be read in process_record. This facilitates independent usage (only dependent on the cache).

File: torcheeg/datasets/module/motor_imagery/strokes.py [hunk 4]
```diff
@@ -0,0 +1,313 @@
...
+from typing import Any, Callable, Dict, Tuple, Union
+from ..base_dataset import BaseDataset
+from ....utils import get_random_dir_path
+from scipy.io import loadmat
+import re
+import pandas as pd
+import mne
+
+
+class StrokePatientsMIDataset(BaseDataset):
```
Comment (tczhangzhi) [inline torcheeg/datasets/module/motor_imagery/strokes.py:11]: Define two datasets: one is StrokePatientsMIDataset, and the other is StrokePatientsMIProcessedDataset. This helps with unit testing and distinguishing different hook behaviors.

File: torcheeg/datasets/module/motor_imagery/strokes.py [hunk 5]
```diff
@@ -0,0 +1,313 @@
...
+            re.findall("sub-(\d\d)_task-motor-imagery_eeg.edf", file)[0])
+        edf_reader = mne.io.read_raw_edf(file, preload=True)
+        epochs = mne.make_fixed_length_epochs(edf_reader,
+                                              duration=8,
+                                              preload=True)
+        data = epochs.get_data(
+        )  # shape(40,33,4000) -num_trial, channels, T—duration（8s）
+
+                  
+        eeg = data[:, :30, :]
```
Comment (tczhangzhi) [inline torcheeg/datasets/module/motor_imagery/strokes.py:157]: before_trial hook?

File: torcheeg/datasets/module/__init__.py [hunk 6]
```diff
@@ -1,4 +1,5 @@
 from .base_dataset import BaseDataset
+from .base_multimodal import BaseMultiModalDataset
```
Comment (tczhangzhi) [inline torcheeg/datasets/module/__init__.py:2]: Please remove it.

**Review by tczhangzhi — CHANGES_REQUESTED** [review]
Thank you for your hard work and contribution. However, I suggest adding unit tests to your implementation to avoid potential errors. Additionally, as a toolkit rather than a personal project, we should consider generality and different scenarios when implementing. A simple tip is to avoid breaking changes to accommodate previously considered cases.

File: torcheeg/transforms/any/baseline.py (hunk 2 above)
Comment (tczhangzhi) [reply torcheeg/transforms/any/baseline.py:78]: Due to the complex related components, you can temporarily keep this part of the code. I will make the modifications.

**Comment by tczhangzhi** [comment]
Thanks for your hard working and contribution, I like the implementation.

Overall, it works well. After I complete testing, I will consider merging it or adding comments.

**Pull request merged**